### ISSUE TITLE
GNS3 - Create SSH functionality for Network-Automation

### ISSUE THREAD

**Issue by richardbui467** [richardbui467/GNS3-Cisco-Automation-Lab#7]
I find that working inside vi or nano inside my Network-Automation container is not as nice as coding in VSCode. Although if I want to do it in VSCode, I would have to copy everything and paste it into vi/nano. There's drawbacks though where the indentation isn't consistent between the two and it's a hasssle to copy things over. But this is especially inconvenient when I decide to delete the Network-Automation node and recreate it to troubleshoot issues. I don't want to have to manually create files and copy over contents of everything I've been making every time.

**Comment by richardbui467** [comment]
Spent a lot of time trying to figure out how to get my computer to ping any of the local nodes. I first started watching videos like [David Bombal's](https://www.youtube.com/watch?v=Gpje9PV1j8U&t=329s) and reading the official [GNS3 documentation](https://docs.gns3.com/docs/using-gns3/advanced/connect-gns3-internet/). I got the impression from various sources that I should import the cloud node to my GNS3 VM, and to bridge a loopback adapter if I didn't have an available ethernet interface. Although, it would never end up with me being able to ping my local computer from my control node and vice versa. Here's some of the things I had tried:

- Creating a loopback adapter and configuring it with IP information within the same subnet as my local computer. Then importing the cloud node to my local endpoint and having it use that loopback adapter. I would then take my control node's interface connected to the cloud and configure it with an address within the same subnet as well. I would also try DHCP, but it wouldn't be able to pull any information. 
- The same step above except I would use my local host's ethernet interface.
- Creating a virtual interface within the GNS3 VM settings, setting it to bridged mode, and then having the cloud node use that. 

I don't remember all the details, but those are roughly the steps I tried using. Although, none of them worked until I had done the following:

- Import the cloud node onto the GNS3 VM
- Connect control node's eth2 interface to the cloud's eth1 interface (which is the GNS3 VM's NAT interface)
- Configure control node's eth2 interface with DHCP to pull an address
- Observed the assigned address, then statically configure eth2 to be within the same subnet
- Pinged my local computer's active wi-fi interface from the control node and vice versa, pings were successful

**Comment by richardbui467** [comment]
After achieving connectivity, I had then looked at getting SSH started up. I immediately looked at getting an SSH server installed onto my endpoint, since I thought I only had the client. So, I had installed OpenSSH server and looked into ways to starting up the SSH daemon. First, I tried entering in "systemctl start sshd", but then I got the below error:

"System has not been booted with systemd as init system (PID 1). Can't operate.
Failed to connect to bus: Host is down"

So then another article had me run "/sbin/service ssh status" to verify the daemon was running and "/sbin/service ssh start" to start it. Unfortunately, I was just told that sshd wasn't a service. But then when I was looking at my /etc/ and /usr/bin directories, I magically came up with the idea to just type in "ssh" instead of "sshd", which worked. I think the idea came to me when I couldn't see that "sshd" existed in those file paths. 

The next step I took was to configure a root password, since I figure SSH would ask for one. I did that entering in "passwd root" and answering the prompts to enter my password as followed. I was wondering if I could use an account that wasn't root though, since I would figure it's a better practice. When I tried making a user named cisco with "useradd" though, it oddly didn't make a home directory. Although, I was still able to make a password for cisco then supply credentials to my PuTTY terminal. Something just felt off though and I felt like if I continued to use cisco, I would just run into problems. So for now I'm going to hold off on exploring that area and save it for later when I can get other things going. 

The 1st time I tried SSHing in as root though, it would refuse the connection. I found that I had to make a line in sshd_config saying "PermitRootLogin yes". Afterwards, I was able to SSH in as root.
